nimbus.1
.\" SPDX-FileType: DOCUMENTATION
.\" SPDX-FileCopyrightText: 2022 Anna <cyber@sysrq.in>
.\" SPDX-License-Identifier: BSD-3-Clause
.Dd July 2, 2022
.Dt NIMBUS 1
.Os
.Sh NAME
.Nm nimbus
.Nd a Nim build system
.Sh SYNOPSIS
.Nm
.Op Fl Fl debug
.Op Fl Fl binDir : Ns Ar path
.Op Fl Fl nimbleDir : Ns Ar path
.Op Fl Fl nim : Ns Ar path
.Op Fl Fl url : Ns Ar url
.Op nim opts...
.Ar sourceDir
.Op Ar buildDir
.Sh DESCRIPTION
The
.Nm
meta-build system generates files for the Ninja build system from Nimble projects.
.Pp
The arguments are as follows:
.Bl -tag -width Ds
.It Ar sourceDir
Source directory,
location where package's
.Sy .nimble
file is placed.
.
.It Ar buildDir
Build directory,
location where all generated files should be placed.
If this argument is omitted, the current directory is used instead.
.Pp
Note that the build directory must be different from the source directory.
.Nm
does not support building inside the source directory and attempting to do that leads to an error.
.
.It Fl Fl binDir : Ns Ar path
Set the executable directory, where project's binaries will be installed.
.
.It Fl Fl debug
Show debugging information.
.
.It Fl Fl nimbleDir : Ns Ar path
Set the Nimble directory, where project's sources will be installed.
.
.It Fl Fl nim : Ns Ar path
Set the Nim executable.
.
.It Fl Fl url : Ns Ar url
If this options is set,
.Nm
will generate and install a
.Pa nimblemeta.json
file.
Some packages specify their dependencies using URLs and
.Nm
is unable to find them unless such metadata file exists.
.El
.Pp
All unrecognized flags are passed to the Nim compiler.
.
.Ss NimScript tasks
.Nm
will create a ninja target for each task defined in the
.Pa .nimble
file.
.
.Ss Running tests
If no tasks named
.Qq test
exist,
.Nm
generates a tester script that can be invoked via
.Ql ninja test .
It will compile and run all Nim files in the
.Pa tests
directory beginning with
.Qq t
in their filename.
.Pp
This behavior is compatible with nimble.
.
.Ss Installing
.Nm
generates an installer script that can be invoked via
.Ql ninja install .
It installs both sources and binaries
.Pq if any .
.Pp
If any of the
.Va installDirs , installFiles , installExt
variables are present in the
.Pa .nimble
file,
.Nm
operates in the whitelist mode.
Please refer the to the nimble documentation for more information.
.Sh ENVIRONMENT
.Bl -tag -width DESTDIR
.It Ev DESTDIR
A variable prepended to each installed target file.
.El
.Sh FILES
The
.Nm
meta-build system generates the following files:
.Bl -tag -width nimblemeta.json
.It Pa build.ninja
Build file for a ninja-compatible build system.
.It Pa installer.nims
Script for installing packages.
.It Pa nimblemeta.json
Metadata file that contains package URL.
.It Pa querytool.nims
Script for extracting metadata from
.Pa .nimble
files.
.It Pa tester.nims
Script for running Nimble-style tests.
.El
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
Build and install a project:
.Bd -literal
mkdir build
cd build
nimbus ..
ninja
sudo ninja install
.Ed
.Sh SEE ALSO
.Bl -bullet -width 1n
.It
.Lk https://nim-lang.org/docs/nimc.html "Nim compiler user guide"
.It
.Lk https://github.com/nim-lang/nimble "Nimble package manager"
.It
.Lk https://ninja-build.org/manual.html "Ninja manual"
.El
.Sh AUTHORS
.An -split
.An Anna
.Aq Mt cyber@sysrq.in
.An Nimble Authors


ChangeLog
2023-04-27  Anna  <cyber@sysrq.in>

	* 1.1.0, 1.1.1 tagged

	* ninjasyntax.nim (build): don't escape args automatically

	* nimbus.nim: use a custom build script

	* tests: refine directory structure

	* nimbus.nimble: add clean task

	* Use separate cache directories for each target

2023-04-25  Anna  <cyber@sysrq.in>

	* nimbus.nim: drop support for patched Nim compilers

	* txt2deps: new utility

2022-07-13  Anna  <cyber@sysrq.in>

	* 0.3.0, 0.3.1 tagged

	* nim.cfg: switch to --gc:orc, disable threads

	* nimbleexecutor.nim (toJsonString): query variables in bulk

	* nimbus.nim: support patched Nim compilers with --depfile option; add
	debug output

	* options.nim (initLogger): set default log level to NOTICE

2022-07-04  Anna  <cyber@sysrq.in>

	* 0.2.4, 0.2.5 tagged

	* Implemented better quoting for generated NimScript and ninja.build.

	* nimbleexecutor.nim (query): add debug output

	* packageinfo.nim (initPackageInfo): use custom spawnX-like wrappers
	to avoid spawning too much threads, add debug info

2022-07-02  Anna  <cyber@sysrq.in>

	* 0.2.0, 0.2.1, 0.2.2, 0.2.3 tagged

	* Fixed build for binaries inside subdirectories.

	* Added support for nimblecache in the build dir.

	* packageinfo.nim (initPackageInfo): execute NimScript in parallel

	* testerscript.nim (writeTesterScript): fix for tests that expect to
	be run from the source root

2022-06-29  Anna  <cyber@sysrq.in>

	* 0.1.4 tagged


src/nimbus.nim
# SPDX-FileCopyrightText: 2022-2023 Anna <cyber@sysrq.in>
# SPDX-License-Identifier: BSD-3-Clause

import std/strutils except escape
import std/[logging, os, sequtils, strtabs, strformat]

import nimbs/[common, options]
import nimbs/[ninjasyntax]
import nimbs/[
  dependencyresolver,
  nimbleexecutor,
  packageinfo,
  packagemetadata,
  version
]
import nimbs/[
  builderscript,
  installerscript,
  querytoolscript,
  testerscript
]

proc processDependencies(requires: seq[string], options: Options): seq[string] =
  ## Checks package dependencies and returns list of paths for the Nim compiler,
  ## shell-quoted.
  for req in requires:
    let dep = parseRequires(req)
    case dep.name
    of "nim":
      let nimVer = getNimVersion(options)
      if not withinRange(nimVer, dep.ver):
        raise unsatisfiedDependencyError("Unsatisfied Nim dependency")
      debug("Found suitable Nim version")
    else:
      result.add(dep.getPath(options).quoteShell)

proc application(ninja: File, input, targetName: string;
                 paths: seq[string], options: Options) =
  debug(fmt"[build.ninja] Generating target for application '{targetName}'")

  var vars = newStringTable()
  vars["targetname"] = targetName
  vars["target"] = "$builddir" / targetName.escape(body = true)
  vars["sourcefile"] = input.escape(body = true)
  vars["nimcache"] = escape(options.getNimCacheBaseDir() / targetName,
                            body = true)
  if paths.len != 0:
    vars["paths"] = escape("-p:" & paths.join(" -p:"), body = true)

  let jsonScript = "$nimcache" / targetName.addFileExt("json").escape
  ninja.build([jsonScript],
    rule = "genscript",
    inputs = [input.escape, "$builder"],
    variables = vars
  )
  ninja.newline()

  ninja.build([vars["target"].addFileExt(ExeExt)],
    rule = "jsonscript",
    inputs = [input.escape, "$builder"],
    implicit = [jsonScript],
    variables = vars
  )
  ninja.newline()

proc task(ninja: File, nimsFile, taskName: string) =
  debug(fmt"[build.ninja] Generating target for task '{taskName}'")

  var vars = newStringTable()
  vars["taskname"] = taskName

  ninja.build([taskName.escape],
    rule = "nimbletask",
    inputs = [nimsFile.escape],
    implicit = ["PHONY"],
    variables = vars
  )

proc setup(options: Options) =
  # if true, nimble-style tests are used
  var nimbleTests = false

  echo "The nimbus build system"
  echo "Version: " & nimbusVersion
  echo "Source dir: " & options.getSourceDir()
  echo "Build dir: " & options.getBuildDir()

  if options.getSourceDir() == options.getBuildDir():
    quit("In-source builds are not allowed")

  debug("Writing query tool script")
  let queryTool = open(options.getBuildDir() / queryToolFileName, fmWrite)
  queryTool.writeQueryToolScript()
  queryTool.close()

  debug("Populating package info")
  let pkgInfo = initPackageInfo(options)
  echo "Project name: " & pkgInfo.name
  echo "Project version: " & pkgInfo.version
  echo "Nim compiler: " & options.getNimBin()
  echo ""

  let nimbusPriv = options.getBuildDir() / "nimbus-private"
  createDir(nimbusPriv)

  let depPaths = processDependencies(pkgInfo.requires, options)

  let nimsFile =
    nimbusPriv / splitFile(pkgInfo.nimbleFile).name.changeFileExt("nims")
  debug("Copying nimble file to nimbus-private")
  copyFile(pkgInfo.nimbleFile, nimsFile)
  debug("Querying tasks")
  let tasks = nimsFile.getTasks(options)

  if options.url.len != 0:
    echo "-- Generating nimblemeta.json"
    let nimblemeta = open(options.getBuildDir() / packageMetadataFileName,
                          fmWrite)
    nimblemeta.writeMetaData(options.url)
    nimblemeta.close()

  if pkgInfo.bin.len != 0:
    echo "-- Generating builder script"
    let builder = open(options.getBuildDir() / builderFileName, fmWrite)
    builder.writeBuilderScript(options)
    builder.close()

  if "test" notin tasks and dirExists(options.getSourceDir() / "tests"):
    echo "-- Generating tester script"
    let tester = open(options.getBuildDir() / testerFileName, fmWrite)
    tester.writeTesterScript(options)
    tester.close()
    nimbleTests = true

  echo "-- Generating installer script"
  let installer = open(options.getBuildDir() / installerFileName, fmWrite)
  installer.writeInstallerScript(pkgInfo, options)
  installer.close()

  echo "-- Generating build.ninja"
  let ninja = open(options.getBuildDir() / "build.ninja", fmWrite)

  debug("[build.ninja] Writing comments")
  ninja.comment("This file is autogenerated by the nimbus build system.")
  ninja.comment("Do not edit by hand.")
  ninja.newline()

  debug("[build.ninja] Writing variables")
  ninja.variable("builddir", options.getBuildDir().escape(body = true))
  ninja.variable("nim", options.getNimBin().escape(body = true))
  ninja.variable("nimbus", getAppFilename().escape(body = true))
  ninja.variable("cmdline", options.getCmdLine().escape(body = true))
  ninja.variable("builder", builderFileName.escape(body = true))
  ninja.newline()

  debug("[build.ninja] Generating 'REGENERATE_BUILD' rule")
  ninja.rule("REGENERATE_BUILD",
    command = "$nimbus $cmdline",
    description = "Regenerating build files.",
    pool = "console",
    generator = true)
  ninja.newline()

  debug("[build.ninja] Generating 'nimscript' rule")
  ninja.rule("nimscript",
    command = "$nim --hints:off e $in",
    description = "Executing NimScript file $in",
    pool = "console")
  ninja.newline()

  if pkgInfo.bin.len != 0:
    debug("[build.ninja] Generating 'genscript' rule")
    ninja.rule("genscript",
      command = "$nim --hints:off e $builder" &
                " --genscript -T:$target $paths $sourcefile",
      description = "Generating build script for Nim application $targetname",
      depfile = "$target".addFileExt("d"),
      deps = "gcc",
      pool = "console")
    ninja.newline()

    debug("[build.ninja] Generating 'jsonscript' rule")
    ninja.rule("jsonscript",
      command = "$nim --hints:off e $builder" &
                " --jsonscript -T:$target $sourcefile",
      description = "Compiling Nim application $targetname",
      depfile = "$target".addFileExt("d"),
      deps = "gcc",
      pool = "console")
    ninja.newline()

  if tasks.len != 0:
    # most tasks are supposed to be run from the project root
    debug("[build.ninja] Generating 'nimbletask' rule")
    ninja.rule("nimbletask",
      command = "cd $sourcedir && $nim --hints:off $taskname $in",
      description = "Executing task $taskname",
      pool = "console")
    ninja.newline()

  debug("[build.ninja] Generating 'PHONY' target")
  ninja.comment("Phony build target, always out of date")
  ninja.build(["PHONY"], rule = "phony")
  ninja.newline()

  for taskName in tasks:
    ninja.task(nimsFile, taskName)
    ninja.newline()

  debug("[build.ninja] Generating 'build.ninja' target")
  ninja.build(["build.ninja"],
    rule = "REGENERATE_BUILD",
    inputs = [pkgInfo.nimbleFile.escape])
  ninja.newline()

  debug("[build.ninja] Generating 'reconfigure' target")
  ninja.build(["reconfigure"],
    rule = "REGENERATE_BUILD",
    implicit = ["PHONY"])
  ninja.newline()

  for bin in pkgInfo.bin:
    let input = pkgInfo.getSourceDir(options) / bin.addFileExt("nim")
    let targetName = bin.lastPathPart
    ninja.application(input, targetName, depPaths, options)

  debug("[build.ninja] Generating 'all' target")
  ninja.build(["all"],
    rule = "phony",
    inputs = pkgInfo.bin.mapIt(
      "$builddir" / it.lastPathPart.addFileExt(ExeExt).escape
    )
  )
  ninja.default(["all"])
  ninja.newline()

  if nimbleTests:
    debug("[build.ninja] Generating 'test' target")
    ninja.build(["test"],
      rule = "nimscript",
      inputs = [testerFileName.escape(body = true)],
      implicit = ["PHONY", "all"])
    ninja.newline()

  debug("[build.ninja] Generating 'install' target")
  ninja.build(["install"],
    rule = "nimscript",
    inputs = [installerFileName.escape(body = true)],
    implicit = ["PHONY", "all"])
  ninja.close()

when isMainModule:
  var opt = parseCmdLine()
  if opt.showHelp:
    writeHelp() # quits
  opt.setLogger
  opt.setSourceDir
  opt.setBuildDir
  opt.setNimBin
  opt.setNimbleDir
  opt.setBinDir
  opt.setup()
